sql/parser/eval.go

```diff
@@ -581,6 +605,380 @@ func init() {
 	cmpOps[cmpArgs{In, tupleType, tupleType}] = evalTupleIN
 }
 
+// TypeCheckExpr type-cheks an SQL expression. Expression checking is a
+// mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
```

Comment:
`eval.go` has grown to be a bloated beast. Let's put this in a new `type_check.go` file. I think it is also time to move all of the datum stuff into a `datum.go` file, but let's hold off on doing that until Vivek's date/timestamp change is in.